File: app/src/main/java/com/bitchat/android/nostr/GeohashMessageHandler.java
package com.bitchat.android.nostr;

import android.app.Application;
import android.util.Log;
import com.bitchat.android.model.BitchatMessage;
import com.bitchat.android.ui.ChatState;
import com.bitchat.android.ui.DataManager;
import com.bitchat.android.ui.MessageManager;
import java.util.ArrayDeque;
import java.util.Date;
import java.util.HashSet;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import android.os.Handler;
import android.os.Looper;

public class GeohashMessageHandler {
    private static final String TAG = "GeohashMessageHandler";

    private final Application application;
    private final ChatState state;
    private final MessageManager messageManager;
    private final GeohashRepository repo;
    private final Executor scope;
    private final DataManager dataManager;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    private final ArrayDeque<String> processedIds = new ArrayDeque<>();
    private final HashSet<String> seen = new HashSet<>();
    private final int max = 2000;

    public GeohashMessageHandler(
            Application application,
            ChatState state,
            MessageManager messageManager,
            GeohashRepository repo,
            Executor scope,
            DataManager dataManager
    ) {
        this.application = application;
        this.state = state;
        this.messageManager = messageManager;
        this.repo = repo;
        this.scope = scope;
        this.dataManager = dataManager;
    }

    private boolean dedupe(String id) {
        if (seen.contains(id)) return true;
        seen.add(id);
        processedIds.addLast(id);
        if (processedIds.size() > max) {
            String old = processedIds.removeFirst();
            seen.remove(old);
        }
        return false;
    }

    public void onEvent(NostrEvent event, String subscribedGeohash) {
        scope.execute(() -> {
            try {
                if (event.getKind() != 20000) return;
                String tagGeo = event.getTags().stream()
                        .filter(tag -> tag.size() >= 2 && tag.get(0).equals("g"))
                        .map(tag -> tag.get(1))
                        .findFirst().orElse(null);

                if (tagGeo == null || !tagGeo.equalsIgnoreCase(subscribedGeohash)) return;
                if (dedupe(event.getId())) return;

                PoWPreferenceManager.PoWSettings pow = PoWPreferenceManager.getCurrentSettings();
                if (pow.isEnabled() && pow.getDifficulty() > 0) {
                    if (!NostrProofOfWork.validateDifficulty(event, pow.getDifficulty())) return;
                }

                if (dataManager.isGeohashUserBlocked(event.getPubkey())) return;

                repo.updateParticipant(subscribedGeohash, event.getPubkey(), new Date(event.getCreatedAt() * 1000L));
                event.getTags().stream()
                        .filter(tag -> tag.size() >= 2 && tag.get(0).equals("n"))
                        .findFirst()
                        .ifPresent(tag -> repo.cacheNickname(event.getPubkey(), tag.get(1)));
                event.getTags().stream()
                        .filter(tag -> tag.size() >= 2 && tag.get(0).equals("t") && tag.get(1).equals("teleport"))
                        .findFirst()
                        .ifPresent(tag -> repo.markTeleported(event.getPubkey()));

                try {
                    GeohashAliasRegistry.put("nostr_" + event.getPubkey().substring(0, 16), event.getPubkey());
                } catch (Exception e) {
                    // Ignore
                }

                NostrIdentity my = NostrIdentityBridge.deriveIdentity(subscribedGeohash, application);
                if (my.getPublicKeyHex().equalsIgnoreCase(event.getPubkey())) return;

                boolean isTeleportPresence = event.getTags().stream()
                        .anyMatch(tag -> tag.size() >= 2 && tag.get(0).equals("t") && tag.get(1).equals("teleport"))
                        && event.getContent().trim().isEmpty();
                if (isTeleportPresence) return;

                String senderName = repo.displayNameForNostrPubkeyUI(event.getPubkey());
                boolean hasNonce = false;
                try {
                    hasNonce = NostrProofOfWork.hasNonce(event);
                } catch (Exception e) {
                    // Ignore
                }

                Integer powDifficulty = null;
                if (hasNonce) {
                    try {
                        int difficulty = NostrProofOfWork.calculateDifficulty(event.getId());
                        if (difficulty > 0) {
                            powDifficulty = difficulty;
                        }
                    } catch (Exception e) {
                        // Ignore
                    }
                }

                BitchatMessage msg = new BitchatMessage(
                        event.getId(),
                        senderName,
                        event.getContent(),
                        new Date(event.getCreatedAt() * 1000L),
                        false,
                        repo.displayNameForNostrPubkey(event.getPubkey()),
                        "nostr:" + event.getPubkey().substring(0, 8),
                        null,
                        "#" + subscribedGeohash,
                        powDifficulty
                );
                mainHandler.post(() -> messageManager.addChannelMessage("geo:" + subscribedGeohash, msg));
            } catch (Exception e) {
                Log.e(TAG, "onEvent error: " + e.getMessage());
            }
        });
    }
}

File: app/src/main/java/com/bitchat/android/nostr/GeohashMessageHandler.kt
package com.bitchat.android.nostr

import android.app.Application
import android.util.Log
import com.bitchat.android.model.BitchatMessage
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.launch
import java.util.Date

/**
 * GeohashMessageHandler
 * - Processes kind=20000 Nostr events for geohash channels
 * - Updates repository for participants + nicknames
 * - Emits messages to MessageManager
 */
class GeohashMessageHandler(
    private val application: Application,
    private val repo: GeohashRepository,
    private val dataManager: com.bitchat.android.ui.DataManager,
    private val addChannelMessage: (String, BitchatMessage) -> Unit
) {
    companion object { private const val TAG = "GeohashMessageHandler" }

    // Simple event deduplication
    private val processedIds = ArrayDeque<String>()
    private val seen = HashSet<String>()
    private val max = 2000

    @Synchronized
    private fun dedupe(id: String): Boolean {
        if (seen.contains(id)) return true
        seen.add(id)
        processedIds.addLast(id)
        if (processedIds.size > max) {
            val old = processedIds.removeFirst()
            seen.remove(old)
        }
        return false
    }

    @Synchronized
    internal fun clearAccountState() {
        processedIds.clear()
        seen.clear()
    }

    internal fun onEvent(
        event: NostrEvent,
        subscribedGeohash: String,
        accountEpoch: NostrAccountEpoch
    ) {
        val accountContext =
            NostrInboundAccountLifecycle.contextFor(accountEpoch)
                ?: return
        accountContext.receiveScope.launch {
            try {
                if (!NostrInboundAccountLifecycle.isCurrent(accountEpoch)) {
                    return@launch
                }
                if (event.kind != NostrKind.EPHEMERAL_EVENT && event.kind != NostrKind.GEOHASH_PRESENCE) return@launch
                val tagGeo = event.tags.firstOrNull { it.size >= 2 && it[0] == "g" }?.getOrNull(1)
                if (tagGeo == null || !tagGeo.equals(subscribedGeohash, true)) return@launch

                // PoW validation (if enabled) - apply to chat messages primarily
                if (event.kind == NostrKind.EPHEMERAL_EVENT) {
                    val pow = PoWPreferenceManager.getCurrentSettings()
                    if (pow.enabled && pow.difficulty > 0) {
                        if (!NostrProofOfWork.validateDifficulty(event, pow.difficulty)) return@launch
                    }
                }

                // Normalize pubkey to lowercase for consistent blocking and storage
                val pubkey = event.pubkey.lowercase()

                // Blocked users check (use injected DataManager which has loaded state)
                if (dataManager.isGeohashUserBlocked(pubkey)) return@launch

                val isTeleportPresence = event.tags.any { it.size >= 2 && it[0] == "t" && it[1] == "teleport" } &&
                                         event.content.trim().isEmpty()
                NostrInboundAccountLifecycle.runIfCurrent(accountEpoch) {
                    if (dedupe(event.id)) return@runIfCurrent
                    repo.updateParticipant(
                        subscribedGeohash,
                        pubkey,
                        Date(event.createdAt * 1000L)
                    )
                    event.tags.find {
                        it.size >= 2 && it[0] == "n"
                    }?.let {
                        repo.cacheNickname(pubkey, it[1])
                    }
                    event.tags.find {
                        it.size >= 2 && it[0] == "t" && it[1] == "teleport"
                    }?.let {
                        repo.markTeleported(pubkey)
                    }
                    GeohashAliasRegistry.put(
                        "nostr_${pubkey.take(16)}",
                        pubkey
                    )
                    val emitMessage =
                        event.kind != NostrKind.GEOHASH_PRESENCE &&
                            !isTeleportPresence &&
                            !NostrIdentityBridge.deriveIdentity(
                                subscribedGeohash,
                                application
                            ).publicKeyHex.equals(pubkey, true)
                    if (emitMessage) {
                        val hasNonce = try {
                            NostrProofOfWork.hasNonce(event)
                        } catch (_: Exception) {
                            false
                        }
                        val msg = BitchatMessage(
                            id = event.id,
                            sender = repo.displayNameForNostrPubkeyUI(pubkey),
                            content = event.content,
                            timestamp = Date(event.createdAt * 1000L),
                            isRelay = false,
                            originalSender = repo.displayNameForNostrPubkey(pubkey),
                            senderPeerID = "nostr:${pubkey.take(8)}",
                            senderNostrPubkey = pubkey,
                            mentions = null,
                            channel = "#$subscribedGeohash",
                            powDifficulty = try {
                                if (hasNonce) {
                                    NostrProofOfWork.calculateDifficulty(event.id)
                                        .takeIf { it > 0 }
                                } else {
                                    null
                                }
                            } catch (_: Exception) {
                                null
                            }
                        )
                        addChannelMessage(
                            "geo:$subscribedGeohash",
                            msg
                        )
                    }
                }
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                Log.e(TAG, "onEvent error: ${e.message}")
            }
        }
    }
}
